disorder-corpus/src/Disorder/Corpus.hs

```diff
@@ -1,19 +1,82 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE OverloadedStrings #-}
 module Disorder.Corpus (
-    cooking
+    Cooking(..)
+  , cooking
+
+  , Muppet(..)
   , muppets
+
+  , Southpark(..)
   , southpark
+
+  , Simpson(..)
   , simpsons
+
+  , Virus(..)
   , viruses
+
+  , Colour(..)
   , colours
+
+  , Weather(..)
   , weather
+
+  , Water(..)
   , waters
+
+  , Boat(..)
   , boats
   ) where
 
-import           Data.Text
+import           Data.Data (Data)
+import           Data.Eq (Eq)
+import           Data.Functor ((<$>))
+import qualified Data.List as List
+import           Data.Ord (Ord)
+import           Data.String (IsString(..))
+import           Data.Typeable (Typeable)
+
+import           GHC.Generics (Generic)
+
+import           Test.QuickCheck (Arbitrary(..), Gen, oneof, elements)
 
-cooking :: [Text]
+import           Text.Show (Show)
+import           Text.Read (Read)
+
+
+-- | Generate something in the corpus or something completely bonkers.
+genCorpus :: Arbitrary b => (a -> b) -> [a] -> Gen a -> Gen b
+genCorpus f corpus gen =
+  oneof [
+      f <$> elements corpus
+    , f <$> gen
+    ]
+
+-- | Shrinks a 'b' to something in the corpus. If 'b' is already from the
+--   corpus then shrink to nothing.
+shrinkCorpus :: Eq a => (a -> b) -> (b -> a) -> [a] -> b -> [b]
+shrinkCorpus f g corpus x =
+  if List.elem (g x) corpus then
+    []
+  else
+    f <$> corpus
+
+
+newtype Cooking a =
+  Cooking {
+      unCooking :: a
+    } deriving (Eq, Ord, Read, Show, Generic, Data, Typeable)
+
+instance (Eq a, IsString a, Arbitrary a) => Arbitrary (Cooking a) where
+  arbitrary =
+    genCorpus Cooking cooking arbitrary
+  shrink =
+    shrinkCorpus Cooking unCooking cooking
+
+cooking :: IsString a => [a]
```

Comment:
Yeh that is fine. There are already a few places that use it for FilePath that would be helped by this.